tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,175 @@
+#
+valueListT = BaseCppType('at', 'ValueList') # TODO, not sure this type exists
+valueT = BaseCppType('ir', 'Value')
+intArrayT = BaseCppType('std', 'vector<int64_t>') # TODO this should probably be different
+
+def process_ir_types(func: FunctionSchema) -> Tuple[List[NamedCType], List[NamedCType], List[NamedCType]]:
+    """
+    We handle Tensor arguments specially as 'IR Values', and everything else (?) as usual.
+
+    TODO, a less awful way of achieving this than what I have done, which is basically to reimplement
+    half of argumenttype_type and fall back to it for non tensors.
+    """
+    types = []
+    value_types = []
+    scalar_types = []
+    for arg in func.schema_order_arguments():
+        t = arg.type
+        mutable = arg.is_write
+        binds = arg.name
+        if isinstance(t, BaseType):
+            if t.name == BaseTy.Tensor:
+                if mutable and not local.use_const_ref_for_mutable_tensors():
+                    types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))
+                    value_types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))
+                else:
+                    types.append(NamedCType(binds, ConstRefCType(BaseCType(valueT))))
+                    value_types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        elif isinstance(t, OptionalType):
+            if str(t.elem) == 'Tensor':
+                if mutable and not local.use_const_ref_for_mutable_tensors():
+                    types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))  # TODO: fix this discrepancy
+                    value_types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))  # TODO: fix this discrepancy
+                else:
+                    types.append(NamedCType(binds, ConstRefCType(OptionalCType(BaseCType(valueT)))))
+                    value_types.append(NamedCType(binds, ConstRefCType(OptionalCType(BaseCType(valueT)))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        elif isinstance(t, ListType):
+            if str(t.elem) == 'Tensor':
+                types.append(NamedCType(binds, BaseCType(valueListT)))
+                value_types.append(NamedCType(binds, BaseCType(valueListT)))
+            elif str(t.elem) == 'Tensor?':
+                types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+                value_types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+            elif str(t.elem) == 'int':
+                types.append(NamedCType(binds, BaseCType(intArrayT)))
```

Comment:
@ezyang What type should I use here?  I used a std::vector<int64_t> for now, but I was wondering if I should use at::Array or some other PyTorch type.

Reply:
Can't use std::array/at::Array because you don't know what the size of the vector is. std::vector is fine. Maybe also consider SmallVector/DimVector